cosmos/ethermint
# implement ethereum-compatible websockets server

Closes: #355

## Description



- [x] start a websockets server when the rest-server starts 
- [x] begin ethereum pubsub api (eth_subscribe, eth_unsubscribe)
- [x] allow subscriptions for `newHeads`
- [x] allow subscriptions for `logs`
- [x] allow subscriptions for `newPendingTransactions`
- [x] stub subscriptions for `syncing`

test instructions:
```
make install
./init.sh
``` 

in another terminal: (can optionally specify  `--wsport [port]` default is 8546)
```
emintcli rest-server --laddr "tcp://localhost:8545" --unlock-key mykey --chain-id 8 --trace
```

in another terminal, install wscat or use whichever ws client you like:
```
npm i -g wscat
wscat -c "ws://localhost:8546"
```

the usual eth api is supported:

```
> {"jsonrpc":"2.0","method":"eth_blockNumber","params":[],"id":1}
< {"id":1,"jsonrpc":"2.0","result":"0x2"}
```
for block subscriptions:
```
> {"id": 1, "method": "eth_subscribe", "params": ["newHeads", {}]}
< {"jsonrpc":"2.0","result":"0x44e010cb2c3161e9c02207ff172166ef","id":1}
```
you should see new block headers every couple seconds

for logs:
```
> {"id": 1, "method": "eth_subscribe", "params": ["logs"]}
< {"jsonrpc":"2.0","result":"0x4f2f0f10e345e3f5ddaa8da3d8eafc6f","id":1}
< {"jsonrpc":"2.0","method":"eth_subscription","params":{"subscription":"0x4f2f0f10e345e3f5ddaa8da3d8eafc6f","result":[{"address":"0xf025c23e267a7cba265e8baf627dd2cb1d010a44","topics":["0x775a94827b8fd9b519d36cd827093c664f93347070a554f65e4a6f56cd738898","0x0000000000000000000000000000000000000000000000000000000000000011"],"data":"0x","blockNumber":"0x1a","transactionHash":"0xcf07c0ff164121b2984209be151fee6fa2734fd11aca531240249e4dc80e5ecc","transactionIndex":"0x0","blockHash":"0x0000000000000000000000000000000000000000000000000000000000000000","logIndex":"0x0","removed":false}]}}
```
in another terminal send a tx that emits some logs: `MODE=rpc go test ./tests/ -run TestEth_GetTransactionLogs`
 you should see the logs in the ws console

can also handle logs subscription with criteria:
```
>> {"id": 1, "method": "eth_subscribe", "params": ["logs", {"address": "0x8320fe7702b96808f7bbc0d4a888ed1468216cfd", "topics": ["0xd78a0cb8bb633d06981248b816e7bd33c2a35a6089241d099fa519e361cab902"]}]}
<< {"jsonrpc":"2.0","id":2,"result":"0x4a8a4c0517381924f9838102c5a4dcb7"}
```

for pending txs:
```
> {"id": 1, "method": "eth_subscribe", "params": ["newPendingTransactions"]}
< {"jsonrpc":"2.0","result":"0x7127b43625439e8e92fd9f653fd490eb","id":1}
```
re-run step to send a tx, you should see the transaction in the ws console
```
< {"jsonrpc":"2.0","method":"eth_subscription","params":{"subscription":"0x7127b43625439e8e92fd9f653fd490eb","result":"0x23d4907a91f121f9418898ede340711978e714b4fe17ed7ae55b978c0c889284"}}
```
______

For contributor use:


______

For admin use:
